src/ascii_str.rs

```diff
@@ -651,40 +700,65 @@ impl<'a> Iterator for Lines<'a> {
             .chars()
             .position(|chr| chr == AsciiChar::LineFeed)
         {
-            let line = if idx > 0 && self.string[idx - 1] == AsciiChar::CarriageReturn {
-                &self.string[..idx - 1]
+            // SAFETY: `idx` is guaranteed to be `1..len`, as we get it from `position` as `0..len` and make sure it's not `0`.
+            let line = if idx > 0
+                && *unsafe { self.string.as_slice().get_unchecked(idx - 1) }
+                    == AsciiChar::CarriageReturn
+            {
+                // SAFETY: As per above, `idx` is guaranteed to be `1..len`
+                unsafe { self.string.as_slice().get_unchecked(..idx - 1).into() }
             } else {
-                &self.string[..idx]
+                // SAFETY: As per above, `idx` is guaranteed to be `0..len`
+                unsafe { self.string.as_slice().get_unchecked(..idx).into() }
             };
-            self.string = &self.string[idx + 1..];
+            // SAFETY: As per above, `idx` is guaranteed to be `0..len`, so at the extreme, slicing `len..` is a valid empty slice.
+            self.string = unsafe { self.string.as_slice().get_unchecked(idx + 1..).into() };
             Some(line)
         } else if self.string.is_empty() {
             None
         } else {
             let line = self.string;
-            self.string = &self.string[..0];
+            // SAFETY: Slicing `..0` is always valid and yields an empty slice
+            self.string = unsafe { AsciiStr::from_ascii_unchecked(b"") };
             Some(line)
         }
     }
 }
+
 impl<'a> DoubleEndedIterator for Lines<'a> {
     fn next_back(&mut self) -> Option<&'a AsciiStr> {
         if self.string.is_empty() {
             return None;
         }
-        let mut i = self.string.len();
-        if self.string[i - 1] == AsciiChar::LineFeed {
-            i -= 1;
-            if i > 0 && self.string[i - 1] == AsciiChar::CarriageReturn {
-                i -= 1;
-            }
-        }
-        self.string = &self.string[..i];
-        while i > 0 && self.string[i - 1] != AsciiChar::LineFeed {
-            i -= 1;
+
+        // If we end with `LF` / `CR/LF`, remove them
+        if let [slice @ .., AsciiChar::CarriageReturn, AsciiChar::LineFeed]
```

Comment:
Subslice pattern were stabilized with Rust 1.42.0 but nevertheless I don't want to update to this version because Debian only ships 1.41.1. Do you want to change this, for example using `self.string.as_bytes().last()` or something?

Reply:
Yes, what I'm currently thinking of doing is similar to the original approach, check if `self.string.last()` is `LF`, and if so, decrease it's len by 1, then check for `CR` within the if above and if so, decrease it by 1 again.

Reply:
Ah, I believe I found a bug in my implementation, If we had a string that ended in [.., LF, LF], the invariant that `lf_rev_pos != 0` is broken.
The old implementation seemed to skip empty lines, which would fix the current implementation too by turning the `if` into a `while`, but I don't think that's what the back iterator should do.
There don't seem to be any tests for multiple consecutive newlines regarding the back iterator, like there are for the forward iterator, so I'll attempt to make an implementation that works with `next()` and add some tests for it.

Unfortunately, I might not be able to do finish it today, but as soon as I'm able to finish it I'll push it.

Reply:
I was mistaken, the fact that `lf_rev_pos` has to be bigger than 0 was not necessary, the function worked fine even when `lf_rev_pos == 0`, it's just that the string returned would be empty, but that's what `Lines` is supposed to do, sorry.
I've fixed the issues regarding compatibility for "1.41" and extended the test for the back iterator.